mypy/checkexpr.py

```diff
@@ -1773,6 +1775,28 @@ def infer_function_type_arguments(
                     callee_type, args, arg_kinds, formal_to_actual, inferred_args, context
                 )
 
+            return_type = get_proper_type(callee_type.ret_type)
+            if isinstance(return_type, CallableType):
+                # fixup:
+                # def [T] () -> def (T) -> T
+                # into
+                # def () -> def [T] (T) -> T
+                for i, argument in enumerate(inferred_args):
+                    if isinstance(get_proper_type(argument), UninhabitedType):
+                        # un-"freshen" the type variable :^)
```

Comment:
I don't like this. A type `def [T] () -> <anything with T>` doesn't make sense, and we should figure out why it was inferred/constructed in the first place. This is really just sweeping dust under the carpet.

Reply:
I can try to explain why this happens:

Have a function that takes a `Callable[Concatenate[T, P], ...]` and then returns a `Callable[P, Callable[[T], ...]]`. This could be some utility function that curries (in reverse), for example.

Now, the problem is that if you use this on `def [T] (T, int) -> ...` then this will turn into... `def [T] (int) -> (T) -> ...`. At least, as currently implemented. Thus, I see two solutions:

- Fix the new function at call time (what I do here) when we have all this information
- Fix the function at construction time, which I'm not sure how to do: I'm not quite sure how to scan through the parameters to find if type variables are still used. For example, this should still work: `def [T] (T, T) -> ...` into `def [T] (T) -> T -> ...`.

Sorry if this is confusing!

Reply:
Ah just to be clear, the problem with `def [T] (int) -> (T) -> ...` is we'll infer an `UninhabitedType` for `T`, meaning that we can't actually use the function anymore.

Reply:
Interesting. I am actually working on a (tangentially related) PR that will likely fix this (or at least will allow a more principled solution). I will post a link here when it is ready.